perf: make screen size and position calculations more efficient

vanaigr opened the pull request:
NOTE: all referenced commits have been shifted by one, first commit now contains benchmarks.

### 1. Breakindent and strcmp()
`win_lbr_chartabsize()` recomputes `breakindent` of a line up to 2 times in order to determine the size of the character. [The function for computing breakindent](https://github.com/neovim/neovim/blob/0a598c13b1863ba1aff378027c4376e3ab7048ee/src/nvim/indent.c#L785) caches the last result, but as part of checking if the cached value could be used it strcmp's the given line to the cached (copied) line. Since `win_lbr_chartabsize()` is often called in a loop and none of the parameters that affect `breakindent` change, this results in strcmp being called for 2 identical lines (worst case for strcmp) for every character in the line starting from the first wrapped row (even though the value is almost never affects the final size of that character).

The first commit fixes the ussue of repeatedly comparing the lines by caching the indentation value in the `chartabsize_T` struct. I also decided to preserve the size of the structure by removing `cts_has_virt_text` and changing `cts_row` to `virt_row` since it is used only for virtual text and can be set to negative value if the line does not contain virtual text.

### 2. Unnecessary loop
`getvcol()` computes virtual column until the last position by iterating until the current position reaches the first byte of the character containing the last position. To compute the first byte of the character, it uses an extra loop at the start to find if the position is past the end of the line and sets it to the end of the line before computing the index of its first byte.

The second commit substitutes comparison with the first byte of the last position with checking if next position is greater than the last position, which allows removing the loop at the start.

### 3. Non-ascii characters are decoded multiple times.
In `getvcol()`, when the size of a Unicode character is computed, the character is decoded and validated (all bytes are read again) once in [`utf_ptr2cells()`](https://github.com/neovim/neovim/blob/0a598c13b1863ba1aff378027c4376e3ab7048ee/src/nvim/mbyte.c#L512). Then, when the pointer is advanced to the next character, its length is computed again (by reading all its bytes) in [`utfc_ptr2len()`](https://github.com/neovim/neovim/blob/0a598c13b1863ba1aff378027c4376e3ab7048ee/src/nvim/mbyte.c#L848), and then it is (possibly) decoded again in [`utf_composinglike()`](https://github.com/neovim/neovim/blob/0a598c13b1863ba1aff378027c4376e3ab7048ee/src/nvim/mbyte.c#L712). The next character is decoded once in `utf_composinglike()`, then its length is computed in `utfc_ptr2len()`, then it is (possibly) decoded a second time if it is a composing character, or it goes though all the steps the first character went. Typically a single Unicode character will be decoded twice and its length will be computed twice per one iteration.
Additionally, the function for decoding a character ([`utf_ptr2char()`](https://github.com/neovim/neovim/blob/0a598c13b1863ba1aff378027c4376e3ab7048ee/src/nvim/mbyte.c#L598)) is and'ing all bytes with masks like `0x3f` to remove the utf8 `10` and `11` bits, but since we know their values, we can subtract these bits instead of and'ing them with masks, which makes it possible to combine all the subtracted bits into one constant.

The issue of duplicate decoding and checking is fixed by adding 3 new functions: `utf_ptr2CharInfo()`, `ptr2CharInfo()` and `utfc_next()`, and modifying `getvcol()`. The first function decodes a character like `utf_ptr2char()`, but this new function returns more information about the decoded character (like its length and whether it is valid). The second function uses the first function to decode a multibyte character, but does a quick return if the character is single byte. The third function works like `utfc_ptr2len()`, but reuses previously decoded characters and also returns the last decoded character which is used on the next iteration of 'getvcol()` loop. The second and third functions are always inline because majority of the time they will be used on ascii characters and not inlining them would add overhead of a function call for the most common usecase.

I verified that `ptr2CharInfo()` and the new version of `utf_ptr2char()` return correct code points for all characters 1-4 bytes long (and some 5 and 6 byte characters) and return a negative value/first byte otherwise.

### 4. Optimizing other places that compute text size.
If wrapping is enabled, `screenpos()` will compute the screen size of all lines above the current line (in [`plines_win_nofold()`](https://github.com/neovim/neovim/blob/bbd5c6363c25e8fbbfb962f8f6c5ea1800d431ca/src/nvim/plines.c#L796)), but only `getvcol()` has the optimized loop for computing character sizes. This function and many other have to use either `win_lbr_chartabsize()` or `lbr_chartabsize()` to compute text size or position. Although both functions have a fast path and check for the same\* conditions as `getvcol()`, all of the checks have to be performed on each iteration. Additionally, Unicode characters are decoded multiple times, like in `getvcol()`.

The fourth commit splits the character size functions into 'regular' and 'fast' versions (`win_lbr_chartabsize()` -> `charsize_regular()`, `win_nolbr_chartabsize()` ~> `charsize_fast()`, `win_linetabsize_cts()` -> `linesize_regular()`, `linesize_fast()`). Checks for which version to use are computed when 'charsize' argument is initialized.
Other functions inside plines.c also have access to the `int_charsize_fast()` which is always inlined.
Both versions of 'charsize' functions now also take decoded characters as arguments instead of decoding them from the pointer, which allows to reuse decoded characters in all other places.

### Other changes

1. [Condition in `getvcol()`](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/plines.c#L517) originally checked for tab behavior and used `win_lbr_chartabsize()` if tabs were drawn as '^I'. But `win_lbr_chartabsize()` didn't explicitly check for that and called either `win_chartabsize()` or `win_nolbr_chartabsize()`, both of which checked the same tab behavior as `getvcol()`, but in [`== TAB` condition](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/plines.c#L46). My version makes `getvcol()` use fast path regardless of whether tabs have variable width or are displayed as '^I'.
2. [Docs](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/runtime/doc/mbyte.txt#L664) say that 'overlong sequences are not handled specially and displayed like a valid character'. I don't know how relevant this is since many places that I modified didn't originally handle overlong ascii characters as ascii (e.g. [`== TAB` in getvcol](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/plines.c#L534) and other places, [BYTE2LEN](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/plines.c#L550) which would handle overlong '^A' as a double-width character, [`ascii_iswhite(*ptr)`](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/edit.c#L4353)). My version handles them correctly only in some places.
3. Potential bug in [`coladvance2()`](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/cursor.c#L147): the `head` variable should probably be reset on every iteration of the loop, like in [`win_line()`](https://github.com/neovim/neovim/blob/ec7453804421f738955a5f4460f44172318d42ff/src/nvim/drawline.c#L1355). `win_lbr_chartabsize()` might not reset `*headp` if it calls `win_chartabsize()` or `win_nolbr_chartbasize()`, so `*headp` might contain 'head' value of one of the previous characters. This would be the case if wrapping is enabled and a double-width character intersect window edge. Then it would set `head` to 1 and it would stay 1 for all characters after it. 
4. 'win_lbr_chartbasize()` recomputed the same width for tab if was displayed like ^I and was moved by virtual text. I changed the condition to check for special tab behavior.

### Benchmarks

To verify that my changes actually improve performance, I wrote a Lua script that runs `screenpos()` 10'000 times in a loop and measures the difference in time returned by os.clock(). I don't know how GC and JIT affect the measurements, but most results are within 1% of the average for the test. Both the original and modified versions were compiled in release mode with Clang 14 and GCC 11.4 on Ubuntu x86-64 (WSL). Screen size is 224x57, wrapping is enabled.
There are 12 different tests in total: 4 with 10'000 characters in one line (long wrapped line), 4 with 100 lines with 220 characters each (long lines, none are wrapped), and 4 with 100 lines with 440 characters each (each line takes 2 screen rows). The results are measured for ascii 'i', and for 2 byte 'À', with and without `breakindent`. Time is in microseconds.
`screenpos()` was computed for the end of the last line, the file was scrolled to the end. 
I also did 4 additional test that measured the time it takes to insert 10'000 characters (with `vim.cmd"normal! 10000ix"`) for the same 2 characters with and without `breakindent`. Time is in **milliseconds**.

GCC. Results with `nobreakindent`:
||insert 10000 i|insert 10000 À|10'000 i|10'000 À|220 i| 220 À|440 i| 440 À|
|-|-|-|-|-|-|-|-|-|
|original|1292|6165|54|384|189|555|184|541|
|this|640|2385|41|160|58|204|55|198|

GCC. Results with `breakindent`:
||insert 10000 i|insert 10000 À|10'000 i|10'000 À|220 i| 220 À|440 i| 440 À|
|-|-|-|-|-|-|-|-|-|
|original|44673|99308|4314|10039|343|767|679|1149|
|this|4129|7604|276|559|341|608|344|600|

Clang, `nobreakindent`:
||insert 10000 i|insert 10000 À|10'000 i|10'000 À|220 i| 220 À|440 i| 440 À|
|-|-|-|-|-|-|-|-|-|
|original|1371|4145|58|264|203|453|195|373|
|this|630|2068|37|137|62|180|63|175|

Clang, `breakindent`:
||insert 10000 i|insert 10000 À|10'000 i|10'000 À|220 i| 220 À|440 i| 440 À|
|-|-|-|-|-|-|-|-|-|
|original|46048|94520|4324|10467|464|705|761|1181|
|this|4059|5666|264|383|330|448|360|494|

vanaigr committed:
perf: cache breakindent/showbreak width in win_lbr_chartabsize

breakindent was recomputed on every call to win_lbr_charbabsize() when the character
is past the end of the first row of a wrapped line. Even though the function for computing
breakindent cached the last result, reusing the cached value required strcmp of the cached line with the given line.

vanaigr committed:
perf: remove loop for computing last position in getvcol()

The function computed the start of the multibyte character and iterated
until the current character reached this position. The loop at the start of the function
handed the case where the index passed to the function was past the end of the line.
This loop can be removed if the function instead compares the next position against
the end position end exits the loop if it is greater than the end position.

vanaigr committed:
perf: don't decode utf8 character multiple times in getvcol()

The optimized virtual column calculation loop in getvcol()
was decoding the current character twice: once in ptr2cells()
and the second time in utfc_ptr2len(). For combining charcters, they were
decoded up to 2 times in utfc_ptr2len(). Additionally, the function used to
decode the character could be further optimised.

vanaigr committed:
perf: reuse fast character size calculation algorithm from getvcol()

VanaIgr commented:
I was looking at how to further optimize `win_lbr_chartabsize()` and it looks like some combinations of virtual text, double-width charaters and breakindent don't work as expected. For example, these 4 lines should probably end at the same virtual column:
![image](https://github.com/neovim/neovim/assets/65824523/9ef26c0a-179c-4920-8001-efd78604a8a3)
(white is regular text, red is virtual text. '?' is a double-width character. Virtual text in the third line contains the double-width char even though it is not displayed as red).

I assume that this is what's happening:
The first line's end column is different because ['>' character is counted](https://github.com/neovim/neovim/blob/6f6abd8c02f66a71f1ad46c4596cadd5709898f4/src/nvim/plines.c#L252) before applying breakindent, and the double-width character is not yet considered to be in the last column.
The third line's end column is different because the width of virtual text is [calculated without considering wrapping](https://github.com/neovim/neovim/blob/6f6abd8c02f66a71f1ad46c4596cadd5709898f4/src/nvim/api/extmark.c#L1160), and width of '>' for a double-width character in the last column is not added.

zeertzjq commented:
> The first line's end column is different because ['>' character is counted](https://github.com/neovim/neovim/blob/6f6abd8c02f66a71f1ad46c4596cadd5709898f4/src/nvim/plines.c#L252) before applying breakindent

Are you referring to the 'breakindent' on the second screen line, or the third?

VanaIgr commented:
First line in the file, second screen row.

Window width is 100. `breakindentopt=shift:2`, so it would add 2 cells to the next row of a line.

Virtual text will have a width of 197 cells. The first real character after virtual text is 2 cells. Total size is 199.

After computing total size, the function checks if `vcol + size - 2` is in last of its row. 0 + 199 - 2 is 197. Virtual column 197 is column 97 of a second row. Last column is 99, so head and size are not incremented.

clason commented:
Impressive work! One small comment:

>  measures the difference in time returned by os.clock(). 

That function is not safe for (fast) benchmarking; prefer `vim.uv.hrtime()`, which has a higher resolution.

bfredl commented on src/nvim/mbyte.h:
```diff
@@ -59,3 +62,62 @@ static inline int mb_strcmp_ic(bool ic, const char *s1, const char *s2)
+///
+/// @return information abouth the character. When the sequence is illegal,
+/// 'value' is negative, 'len' is 1.
+static inline CharInfo ptr2CharInfo(char const *const p_in)
```
this publicly visible function should be called `utf_ptr2CharInfo`. The internal helper function should be named something which indicated that (`utf_ptr2CharInfo_impl` is a pattern we have used elsewhere for implementation functions which are not meant to be called directly ).

bfredl commented on src/nvim/mbyte_defs.h:
```diff
@@ -55,3 +55,13 @@ typedef struct {
 } vimconv_T;
+
+typedef struct {
+  int value;  /// code point
+  int len;  /// length in bytes
```
normally we don't require typed ints for these, but as there is a constraint this struct has to be equal in representation to a `uint64_t`, maybe we should use two `u/int32_t` in this case. And/or perhaps write some comment about this.

VanaIgr commented:
> That function is not safe for (fast) benchmarking; prefer `vim.uv.hrtime()`, which has a higher resolution.

I had noticed that the scripts in 'test/benchmarks' used libuv's `hrtime()` instead of 'os.clock()' and ran some of my benchmarks with it to see if there was any significant difference, but the results were largely the same (average with `hrtime()` had around 10 decimal digits while `os.clock()` only had 1-4, but the results were within 1us of each other). 

Since I ran my benchmarks manually and couldn't find a way to copy anything from within neovim to the Windows clipboard, I used `os.clock()` to save some time on typing the results.

> And maybe consider adding these tests to `test/benchmark`, if possible?

I will write proper benchmarks and make them the new first commit.

VanaIgr commented:
Do the optional s390x checks test if Neovim works on a big endian architecture?
If yes, than they are failing probably because `utf_ptr2CharInfo()` assumes the system is little-endian.

dundargoc commented:
> Do the optional s390x checks test if Neovim works on a big endian architecture?

Yes.

clason commented:
> If yes, than they are failing probably because utf_ptr2CharInfo() assumes the system is little-endian.

And yes. All failures are indeed relevant.

vanaigr committed:
test: add screenpos() benchmarks

bfredl commented:
> Is there a simpler way to execute Lua code inside Neovim and return the results to the benchmarks?

Yes, use `exec_lua` from helpers module.

VanaIgr commented:
I added the benchmarks.

VanaIgr commented:
I also added `plines.h` to clint.py `check_includes_ignore` list, since it uses functions from `mbyte.h` to define `init_chartabsize()` and `adv_chartabsize()`. 
Since both are convenience functions, I can also inline them and remove the dependency on `mbyte.h`.

VanaIgr commented:
I removed both functions. Now `plines.h` only includes mbyte_defs.

VanaIgr commented:
I also found a couple of issues:

1. Blockwise put doesn't add correct number of spaces for multibyte characters.

The loop that determines the number of spaces to add to the pasted text ([link](https://github.com/neovim/neovim/blob/786089013d33666a2acb5ba86fb34c716a9cab26/src/nvim/ops.c#L3293)) advances the pointer by a single byte. This is not correct for multibyte characters, since all trailing bytes would be counted separately as 4 cell wide '\<xx\>'. This issue is also present in vim.
For example, copying the text at the top (blockwise, without virtualedit) and pasting it before the b's at the bottom while in blockwise visual mode would result in this:
![image](https://github.com/neovim/neovim/assets/65824523/2da78925-6bfc-4a1a-90c1-00a516d3aaa6) ![image](https://github.com/neovim/neovim/assets/65824523/d698bc2a-3bc8-4f60-aa77-6d7f0d22c3de)
This can be fixed by using `MB_PTR_ADV(cts.cts_ptr)` instead of `cts.cts_ptr++`, and changing the loop to iterate until the end of the string.

This still wouldn't work correctly for tabs, since all widths are computed at virtual column 0, but tab width changes depending on its position. I thought there also may be some issues if the tab's width is greater than window width and breakindent in enabled (since the size of breakindent in this case would be added to the total width of the pasted line) but for some reason this doesn't happen.

2. Lisp indenting doesn't use correct line number.

In [2 places](https://github.com/neovim/neovim/blob/786089013d33666a2acb5ba86fb34c716a9cab26/src/nvim/indent.c#L1275) in `get_lisp_indent()` , the 'lnum' argument for `init_chartabsize_arg()` is `(colnr_T)(that - line)` (which computes the index of the current byte, not the line number).
This can affect the results if the line at that position contains virtual text:
![image](https://github.com/neovim/neovim/assets/65824523/c819000f-3696-483c-a12a-7ca7e699e8bb)
![image](https://github.com/neovim/neovim/assets/65824523/f179859e-eb29-46f4-aa22-92dac898ecdb) ![image](https://github.com/neovim/neovim/assets/65824523/fbfef993-7ec7-40b3-81d7-45c0b0d4a5df)

There is also [one place](https://github.com/neovim/neovim/blob/786089013d33666a2acb5ba86fb34c716a9cab26/src/nvim/indent.c#L1270) in this function in which character is assumed to be 1 byte long and 1 cell wide. I don't know if this can cause any issues.

justinmk commented:
Can you mention the user-facing consequences in `:help news`: https://github.com/neovim/neovim/blob/2dc439c672facaeb8e51ef6aa20efc0e7092eee2/runtime/doc/news.txt#L129 (e.g. "performance of 'breakindent', foo, bar, etc., is now 3x faster for long lines, etc, etc"

Also would be ideal if the commit message contains the explanation from the PR description, especially the list of problems addressed.

vanaigr committed:
docs(news): expand the list of performance improvements

When computing on-screen size or position, the size 'breakindent' and 'showbreak' is now cached,
and checks for whether a faster character size function can be used are performed only once at the start.
Multibyte characters are not decodes multiple times anymore, and character decoding functions are more efficient.
Additionally, the amount of trailing spaces for pasted blockwise text is now calculated correctly for multibyte characters.
Internal lisp formatting now doesn't erroneously use inline virtual text from a different line.

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -722,6 +788,15 @@ bool utf_composinglike(const char *p1, const char *p2)
+/// Check if the next character is a composing character when it
+/// comes after the first. For Arabic sometimes "ab" is replaced with "c", which
+/// behaves like a composing character.
+bool composinglike(int32_t const first, int32_t const next)
```
This name is strange without a `utf_` prefix. Maybe name it `utf_char_composinglike()`?

VanaIgr commented:
Is there anything else I should do?

clason commented:
Yes, fixing the endian issue that makes the test fail (see @bfredl's comment above).

bfredl commented:
Hmm, there might be no more s3900 failures than those known to exist on master, which is also 8+3, see https://github.com/neovim/neovim/pull/26950

clason commented:
Oh, right, I should have looked at the details first. Yes, the original (relevant) s390x failures seem to be gone. 🚀

bfredl commented on src/nvim/plines.h:
```diff
@@ -4,25 +4,88 @@
+typedef bool CSType;
+
+enum {
+  CHARSIZE_REGULAR = false,
```
`kCharsizeRegular` and `kCharsizeFast` and no need for initializers.

VanaIgr replied:
I added explicit initializers to show that these constants are used as booleans and not integers. Should I still remove them?

bfredl replied:
They are used as enum values, where the exact representation doesn't matter. these just get co-erced to 0/1 anyway. So yes, they should be removed.

bfredl commented:
> The fourth commit splits the character size functions into 'regular' and 'fast' versions (win_lbr_chartabsize() -> charsize_regular(), win_nolbr_chartabsize() ~> charsize_fast(), win_linetabsize_cts() -> linesize_regular(), linesize_fast()). Checks for which version to use are computed when 'charsize' argument is initialized.

Would it be possible to benchmark how much faster things become by this last step (compared to do the rest of the improvements but not this)? It looks this change is more or less isolated to a separate commit (thus the rest of step 4 is in an earlier). Is there a reason 'linebreak', 'showbreak' and 'breakindent' cannot be done in the "fast" variant of the algorithm? (or maybe this doesn't make sense, because it is just the same algorithm but with deletion of branches, etc?).

The use of these options are quite popular, and it would be great if these cases are fully optimized as well, also avoiding duplicate code would be nice. OTOH, if the `charsize_fast()` branch is significantly faster and its hard do to this with these options, and we thus keep this branching as is for a long time, it would be good to document that perhaps ("when dealing with text with really long lines, turn off 'linebreak', 'showbreak' and 'breakindent' for best performance")

zeertzjq commented:
There is a typo "myltibyte" in the third commit meessage.

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -482,115 +478,56 @@ static int virt_text_cursor_off(chartabsize_T *cts, bool on_NUL)
-    }
-    posptr = ptr + pos->col;
-    posptr -= utf_head_off(line, posptr);
+  uintptr_t last_pos = (uintptr_t)(line + pos->col);
+  if (last_pos < (uintptr_t)line) {
+    last_pos = UINTPTR_MAX;  // unsigned overflow
   }
```
```suggestion
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -482,115 +478,56 @@ static int virt_text_cursor_off(chartabsize_T *cts, bool on_NUL)
-        // character at pos->col
+      char_size = int_charsize_fast(wp, special_tab, vcol, ci.chr.value);
+      StrCharInfo const next = utfc_next(ci);
+      if ((uintptr_t)next.ptr > last_pos) {
         break;
       }
```
```suggestion
      if (next.ptr - line > pos->col) {
        break;
      }
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -482,115 +478,56 @@ static int virt_text_cursor_off(chartabsize_T *cts, bool on_NUL)
-      if ((posptr != NULL) && (cts.cts_ptr >= posptr)) {
-        // character at pos->col
+      StrCharInfo const next = utfc_next(ci);
+      if ((uintptr_t)next.ptr > last_pos) {
         break;
       }
```
```suggestion
      if (next.ptr - line > pos->col) {
        break;
      }
```

zeertzjq reviewed (commented):
I don't think it's a good idea to cast pointers to integers when the same thing can be achieved by subtracting pointers. The following suggestions are for the third commit:

VanaIgr commented:
> Would it be possible to benchmark how much faster things become by this last step (compared to do the rest of the improvements but not this)?

Yes, here are the benchmarks:
||5'000 i|5'000 À|50*100 i|50*100 À|25*200 i|25*200 À|
|-|-|-|-|-|-|-|
|original|26|179|75|209|73|207|
|3rd commit|13|84|74|180|72|177|
|4th commit w/o splitting/inlining|11|79|63|136|60|133|
|4th commit|12|79|16|85|14|82|

('4th commit w/o splitting/inlining' adds unicode functions from the third commit to other places)


> Is there a reason 'linebreak', 'showbreak' and 'breakindent' cannot be done in the "fast" variant of the algorithm? (or maybe this doesn't make sense, because it is just the same algorithm but with deletion of branches, etc?).

From my early measurements, splitting and inlining was the fastest option. I also experimented with only having one version and trying to make it as fast as possible, but it was still slow ('fast combined, inlined' in the table below).
Since now I don't have to run all benchmarks manually, I benchmarked the code where only the regular version is used and where the two versions are split. I used '50*100 i' benchmark since it makes use of `linesize_` functions, and I can just change which function is called.

||time, us|
|-|-|
|combined, not inlined|63|
|combined, inlined|38|
|fast combined, inlined|33|
|split, fast not inlined|28|
|split, fast inlined (current)|16|

(Inlined here refers to the `charsize_` function, not `linesize_`)

> The use of these options are quite popular, and it would be great if these cases are fully optimized as well

I think it is possible to use `charsize_fast()` even if `breakindent` or `showbreak` are enabled. We can compute the size of the line and then add `indent_size * (num_rows-1)` to it. This wouldn't work if a doublewidth character is present or we need an intermediate virtual column for some other computation.

I have some other ideas on how `showbreak` and `breakindent` could be optimized in `charsize_regular()`, but I don't understand what `max_head_vcol` does. 
`linebreak` would be very hard to optimize, since it must know if the **following** word should be wrapped, which requires computing its size (also, it currently doesn't consider inline virtual text).
![image](https://github.com/neovim/neovim/assets/65824523/9ec0aa60-84cf-46ac-9d3f-be6f697d72a1)

zeertzjq commented:
It seems that the performance issue in 'breakindent' was a v0.9 regression from #22506. Nevertheless the implementation of `win_lbr_chartabsize()` has changed significantly since then.

zeertzjq commented:
`max_head_vcol` if used to differentiate the following three situations:
- Getting the virtcol of the start of a character
- Getting the virtcol of where cursor should be placed
- Getting how many cells of inline virtual text to skip because of 'smoothscroll'

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
+  static uint32_t const corrections[] = {
+    (1U << 31) + 0x80U,  // invalid - set invalid bits (safe to add as first 2 bytes
+    (1U << 31) + 0x80U,  // won't affect highest bit in normal ret). 0x80U for errorret
+    -(0x80U + (0xC0U << 6)),  // multybyte - subtract added UTF8 bits (1..10xxx and 10xxx)
```
```suggestion
    -(0x80U + (0xC0U << 6)),  // multibyte - subtract added UTF8 bits (1..10xxx and 10xxx)
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
 }
 
+/// Convert a UTF-8 byte sequence to a character number.
+/// Doesn't hanlde ascii! only multibyte and illegal sequences.
```
```suggestion
/// Doesn't handle ascii! only multibyte and illegal sequences.
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -102,127 +75,80 @@ int linetabsize(win_T *wp, linenr_T lnum)
+  cts->cur_text_width_right = 0;
+  cts->virt_row = -1;
+  cts->indent_width = INT_MIN;
+  cts->special_tab = !wp->w_p_list || wp->w_p_lcs_chars.tab1;
```
`special_tab` doesn't seem an accurate name for this. `use_tabstop` seems better.

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -102,127 +75,80 @@ int linetabsize(win_T *wp, linenr_T lnum)
-
-/// Get the number of characters taken up on the screen indicated by "cts".
+/// Get the number of characters taken up on the screen for the given cts and position.
 /// "cts->cts_cur_text_width_left" and "cts->cts_cur_text_width_right" are set
```
The `cts_` prefix should also be now removed here.

zeertzjq replied:
```suggestion
/// "cts->cur_text_width_left" and "cts->cur_text_width_right" are set
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -388,39 +321,49 @@ int win_lbr_chartabsize(chartabsize_T *cts, int *headp)
-  return n;
+}
+
+/// Like cts_charsize_regular, except it doesn't handle virtual text,
```
```suggestion
/// Like charsize_regular(), except it doesn't handle virtual text,
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -597,45 +667,41 @@ size_t mb_string2cells_len(const char *str, size_t size)
-    return p[0];
+  uint32_t v0 = p[0];
+  if (EXPECT(v0 < 0x80U, true)) {  // Be quick for ASCII.
+fret:
```
Using `goto` with this label looks ugly. How much is this faster than using `return (int)v0` directly?

VanaIgr replied:
Generated code is the same in release mode

zeertzjq replied:
Hmm, then I think it's better to remove this label and use `return (int)v0` directly.

zeertzjq commented on src/nvim/drawline.c:
```diff
@@ -1336,30 +1336,30 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, s
+
+    CharsizeArg arg;
+    CSType type = init_charsize_arg(&arg, wp, lnum, line);
+    arg.max_head_vcol = (int)start_col;
```
Lint failure
```suggestion
    arg.max_head_vcol = start_col;
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
+  FUNC_ATTR_PURE FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+// uint8_t is a reminder for clang to use smaller cmp
+#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) goto errorret
```
consider surrounding this with `do {` and `} while (0)`

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -597,45 +667,40 @@ size_t mb_string2cells_len(const char *str, size_t size)
-  // Illegal value: just return the first byte.
-  return p[0];
+
+#define STEP(i, ret) uint32_t const v##i = p[i]; \
+  if (EXPECT((uint8_t)(v##i & 0xC0U) != 0x80U, false)) return (int)v0; \
+  if (EXPECT(len == i + 1, false)) return (int)(ret)
```
ditto

zeertzjq replied:
It's a bit confusing to have the second macro argument reference a variable declared inside the macro. I think it's better to move the declaration out of the macro and put the two `if` statements between `do {` and `} while (0)`.

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -597,45 +667,40 @@ size_t mb_string2cells_len(const char *str, size_t size)
+       - ((0xF8U << 24) + S(18) + S(12) + S(6) + S(0)));
+  // len == 6
+  uint32_t const v5 = p[5];
+  if (EXPECT((uint8_t)(v5 & 0xc0U) != 0x80U, false)) {
```
```suggestion
  if (EXPECT((uint8_t)(v5 & 0xC0U) != 0x80U, false)) {
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
+  code_point = (code_point << 6) + (cur = p[5]);
+  if ((uint8_t)(cur & 0xC0U) != 0x80U) {
+errorret:;
+    // 8th bit of corr is always 1, int32_t guerantees
```
```suggestion
    // 8th bit of corr is always 1, int32_t guarantees
```

zeertzjq commented on src/nvim/plines.h:
```diff
@@ -4,25 +4,93 @@
+}
+
+static inline int linetabsize_str(char *s)
+  REAL_FATTR_NONNULL_ALL REAL_FATTR_WARN_UNUSED_RESULT REAL_FATTR_ALWAYS_INLINE;
```
```suggestion
  REAL_FATTR_NONNULL_ALL REAL_FATTR_WARN_UNUSED_RESULT REAL_FATTR_ALWAYS_INLINE;

```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
+
+  code_point = (code_point << 6) + (cur = p[5]);
+  if ((uint8_t)(cur & 0xC0U) != 0x80U) {
+errorret:;
+    // 8th bit of corr is always 1, int32_t guerantees
+    // 2's complement. Value is negative
+    int32_t const result = (int32_t)(corr << 24);
+    ASSUME(result < 0);
+    return result;
```
If I understand correctly, when this function returns a negative value, the actual value doesn't really matter, so I think it's less confusing if it just returns `-1`.
```suggestion
errorret:;
    return -1;
```

zeertzjq replied:
If not then at least fix the typo "guerantee"
```suggestion
errorret:;
    // 8th bit of corr is always 1, int32_t guarantees
    // 2's complement. Value is negative
    int32_t const result = (int32_t)(corr << 24);
    ASSUME(result < 0);
    return result;
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
+  FUNC_ATTR_PURE FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+// uint8_t is a reminder for clang to use smaller cmp
+#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) goto errorret
```
```suggestion
#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) return -1
```

zeertzjq replied:
Also consider surrounding this with `do {` and `} while (0)`

VanaIgr replied:
If `corr` is not used in all return paths, the compiler would move its read to the normal return instead of reading the value at the start of the function. This could increase latency if many different memory locations are read between character decoding. This can happen in `utf_char_compisinglike()`, and `charsize_regular()` when `linebreak` is on or there is virtual text.

I wasn't able to measure any significant difference in my benchmarks

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -597,45 +667,40 @@ size_t mb_string2cells_len(const char *str, size_t size)
+#define STEP(i, ret) uint32_t const v##i = p[i]; \
+  if (EXPECT((uint8_t)(v##i & 0xC0U) != 0x80U, false)) return (int)v0; \
+  if (EXPECT(len == i + 1, false)) return (int)(ret)
+#define S(s) ((uint32_t)0x80U << s)
```
```suggestion
#define S(s) ((uint32_t)0x80U << (s))
```

zeertzjq commented on src/nvim/mbyte.h:
```diff
@@ -44,3 +49,73 @@ extern const uint8_t utf8len_tab[256];
+
+static inline StrCharInfo utf_ptr2StrCharInfo(char *ptr)
+{
+  return (StrCharInfo){ ptr, utf_ptr2CharInfo(ptr) };
```
```suggestion
  return (StrCharInfo){ .ptr = ptr, .chr = utf_ptr2CharInfo(ptr) };
```

zeertzjq commented on src/nvim/edit.c:
```diff
@@ -4347,14 +4351,16 @@ static bool ins_tab(void)
+    int32_t tab_v = (uint8_t)(*tab);
+
+    CharsizeArg arg;
+    CSType type = init_charsize_arg(&arg, curwin, 0, tab);
```
`type` as a variable name is too general. I think it better to use `cstype` for all these variables.

zeertzjq commented on src/nvim/indent.c:
```diff
@@ -1292,37 +1288,33 @@ int get_lisp_indent(void)
 
-              cts.cts_vcol += lbr_chartabsize_adv(&cts);
+              amount += cts_charsize(type, amount, that, (uint8_t)(*that), &arg).width;
+              that++;
```
Hmm, shouldn't `utfc_next()` be used here?

VanaIgr replied:
Yes, I should've used `utfc_next()` in this case

zeertzjq commented on src/nvim/ops.c:
```diff
@@ -3250,19 +3246,20 @@ void do_put(int regname, yankreg_T *reg, int dir, int count, int flags)
-      size_t oldlen = strlen(oldp);
-      chartabsize_T cts;
-      init_chartabsize_arg(&cts, curwin, curwin->w_cursor.lnum, 0, oldp, oldp);
+      size_t oldlen;
```
This declaration can be moved to its first assignment below.

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -451,19 +395,70 @@ static bool in_win_border(win_T *wp, colnr_T vcol)
+    vcol += charsize_regular(arg, ci.ptr, vcol, ci.chr.value).width;
+    ci = utfc_next(ci);
+  }
+
```
```suggestion

  // check for inline virtual text after the end of the line
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -451,19 +395,70 @@ static bool in_win_border(win_T *wp, colnr_T vcol)
+int linesize_regular(CharsizeArg *const arg, int vcol, colnr_T const len)
+{
+  char *const line = arg->line;
+  uintptr_t end_pos = (uintptr_t)(line + len);
+  if (end_pos < (uintptr_t)line) {
+    end_pos = UINTPTR_MAX;  // unsigned overflow
+  }
+
+  StrCharInfo ci = utf_ptr2StrCharInfo(line);
+  while ((uintptr_t)ci.ptr < end_pos && *ci.ptr != NUL) {
```
```suggestion
  StrCharInfo ci = utf_ptr2StrCharInfo(line);
  while (ci.ptr - line < len && *ci.ptr != NUL) {
```

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -451,19 +395,70 @@ static bool in_win_border(win_T *wp, colnr_T vcol)
+  bool const use_tabstop = arg->use_tabstop;
+
+  char *const line = arg->line;
+  uintptr_t end_pos = (uintptr_t)(line + len);
+  if (end_pos < (uintptr_t)line) {
+    end_pos = UINTPTR_MAX;  // unsigned overflow
+  }
+
+  StrCharInfo ci = utf_ptr2StrCharInfo(line);
+  while ((uintptr_t)ci.ptr < end_pos && *ci.ptr != NUL) {
```
```suggestion
  StrCharInfo ci = utf_ptr2StrCharInfo(line);
  while (ci.ptr - line < len && *ci.ptr != NUL) {
```

zeertzjq commented on src/nvim/mbyte.c:
```diff
@@ -528,6 +528,69 @@ int utf_ptr2cells(const char *p)
+    -(0x80U + (0x80U << 6) + (0xE0U << 12)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0xF0U << 18)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0x80U << 18) + (0xF8U << 24)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0x80U << 18) + (0x80U << 24) + (0xF4U << 30)),
```
Why is the last one here `0xF4U`, not `0xFCU`?

VanaIgr replied:
`0xF4U` It is not correct, but the `(0xF4U << 30)` part is 0 in either case.

zeertzjq replied:
Hmm, then it should still be changed to `0xFCU` to avoid future mistakes.

VanaIgr replied:
I think it would be better to remove it and add a comment with a description of why there is no need to subtract the bits of the leading byte

zeertzjq replied:
Yes that make sense as well.

VanaIgr commented:
I think [one](https://github.com/neovim/neovim/blob/fa9a85ae468b9df30ae9e5c05a08c0f124e267df/test/old/testdir/test_listlbr.vim#L62) of the `linebreak` tests is not correct.

The `hijklmn` part should be on the first row, since tab after it should stop at the end of the window (tabstop=4, width=20). `'linebreak'` implementation in `win_charsize_regular()` uses the virtual column of the previous character when computing the size of the current character, which results in tab being 1 cell longer, in this case.

I started working on the `linebreak` algorithm in `charsize_regular()` because I found that tabs can still create extra line break at the start of the line:
![image](https://github.com/neovim/neovim/assets/65824523/6cc28039-37cd-449e-84cd-736f799db6f7)
(tab needs to be followed by a space or a tab)

VanaIgr commented:
I was trying to rewrite the algorithm after noticing that #25276 can still happen when tabs are used
![image](https://github.com/neovim/neovim/assets/65824523/49169428-e835-4f52-8b4b-55f5b60a260b)
And kept running into failing tests.

[This test](https://github.com/neovim/neovim/blob/fa9a85ae468b9df30ae9e5c05a08c0f124e267df/test/old/testdir/test_listlbr.vim#L62) and [this test](https://github.com/neovim/neovim/blob/8df7978fe31635ede12f6dae224e6a6c8f1ba882/test/old/testdir/test_breakindent.vim#L792) both have a word followed by a line breaking character that stops at the end of the line, but the first word causes a line break, while the second one doesn't.

In case of the first test, the `hijklmn` part is followed by a tab, tabstop=4, window width=20. Tab stops at the end of the line.
For the second test, the word is `no`, and it is followed by a space that also ends at the end of the line.

I think the issue is that `'linebreak'` algorithm probably uses previous character's virtual column to calculate size of the current character, and that results in inconsistent behavior in the tests. Tab would increase its width to end up at the same column, while space would just take 1 cell and end up 1 cell away from the window's edge.

aktau commented:
> That function is not safe for (fast) benchmarking; prefer vim.uv.hrtime(), which has a higher resolution.

Just a small drive-by comment, as some of my earliest comments (and work) on this project could be the reason for this (mis)understanding.

The best way to minimize the overhead of the measurement itself is to run an operation N times, measure the time, then divide by N (to get average time per operation). If N is large enough, the time spent in the timing function will be a negligible part of the total, and the reduced precision (in the ns/us scale) will likely not be a problem.

Good:

```
uint64_t t1 = now()
for (size_t i := 0; i < iterations; ++i) {
  op();
}
uint64_t t2 = now()
time_per_operation = double(t2 - t1) / double(iterations);
```

Bad:

```
time_per_operation = 0;
for (size_t i := 0; i < iterations; ++i) {
  t1 = now();
  op();
  dur = t1 - now();
  time_per_operation = rolling_avg(time_per_operation, dur);
}
```

The "good" version is how most benchmarking libraries work nowadays. The number of iterations needed for a stable result can generally be determined automatically, example from Go: https://github.com/golang/go/blob/39ec246e739a787375b00acd92c10311863575a2/src/testing/benchmark.go#L275.

marcelbeumer commented:
Thanks so much everyone!
